cpu/atxmega/atxmega_cpu.c

```diff
@@ -0,0 +1,141 @@
+#include "debug.h"
+
+#ifndef CPU_ATXMEGA_CLK_SCALE_INIT
+#define CPU_ATXMEGA_CLK_SCALE_INIT    CPU_ATXMEGA_CLK_SCALE_DIV1
+#endif
+#ifndef CPU_ATXMEGA_BUS_SCALE_INIT
+#define CPU_ATXMEGA_BUS_SCALE_INIT    CPU_ATXMEGA_BUS_SCALE_DIV1_1
+#endif
+
+extern uint8_t mcusr_mirror;
+
+void avr8_reset_cause(void)
+{
+    if (mcusr_mirror & (1 << RST_PORF_bp)) {
+        DEBUG("Power-on reset.\n");
+    }
+    if (mcusr_mirror & (1 << RST_EXTRF_bp)) {
+        DEBUG("External reset!\n");
+    }
+    if (mcusr_mirror & (1 << RST_BORF_bp)) {
+        DEBUG("Brown-out reset!\n");
+    }
+    if (mcusr_mirror & (1 << RST_WDRF_bp)) {
+        DEBUG("Watchdog reset!\n");
+    }
+    if (mcusr_mirror & (1 << RST_PDIRF_bp)) {
+        DEBUG("Programming and Debug Interface reset!\n");
+    }
+    if (mcusr_mirror & (1 << RST_SRF_bp)) {
+        DEBUG("Software reset!\n");
+    }
+    if (mcusr_mirror & (1 << RST_SDRF_bp)) {
+        DEBUG("Spike Detection reset!\n");
+    }
+}
+
+void __attribute__((weak)) clk_init(void)
+{
+    uint8_t *reg = (uint8_t *)&PR.PRGEN;
+    uint8_t i;
+
+    /* Turn off all peripheral clocks that can be turned off. */
+    for (i = 0; i <= 7; i++) {
+        *(reg++) = 0xff;
+    }
+
+    reg = (uint8_t *)&PR.PRGEN;
+    /* Turn on all peripheral clocks that can be turned on. */
+    for (i = 0; i <= 7; i++) {
+        *(reg++) = 0x00;
+    }
+
+    /* XMEGA A3U [DATASHEET] p.23 After reset, the device starts up running
+     * from the 2MHz internal oscillator. The other clock sources, DFLLs
+     * and PLL, are turned off by default.
+     *
+     * Configure clock to 32MHz with calibration
+     * application note AVR1003
+     *
+     * From errata http://www.avrfreaks.net/forum/xmega-dfll-does-it-work
+     * In order to use the automatic runtime calibration for the 2 MHz or
+     * the 32 MHz internal oscillators, the DFLL for both oscillators and
+     * both oscillators has to be enabled for one to work.
+     */
+    OSC.PLLCTRL = 0;
+
+    /* Enable the internal PLL & 32MHz & 32KHz oscillators */
+    OSC.CTRL |= OSC_PLLEN_bm | OSC_RC32MEN_bm | OSC_RC32KEN_bm;
+
+    /* Wait for 32Khz oscillator to stabilize */
+    while (!(OSC.STATUS & OSC_RC32KRDY_bm)) {}
+
+    /* Wait for 32MHz oscillator to stabilize */
+    while (!(OSC.STATUS & OSC_RC32MRDY_bm)) {}
+
+    /* Enable DFLL - defaults to calibrate against internal 32Khz clock */
+    DFLLRC32M.CTRL = DFLL_ENABLE_bm;
+
+    /* Enable DFLL - defaults to calibrate against internal 32Khz clock */
+    DFLLRC2M.CTRL = DFLL_ENABLE_bm;
+
+    atxmega_set_prescaler(CPU_ATXMEGA_CLK_SCALE_INIT,
+                          CPU_ATXMEGA_BUS_SCALE_INIT);
+
+    /* Disable CCP for Protected IO register and set new value*/
+    /* Switch to 32MHz clock */
+    _PROTECTED_WRITE(CLK.CTRL, CLK_SCLKSEL_RC32M_gc);
+
+    /*
+     * Previous instruction takes 3 clk cycles with -Os option
+     * we need another clk cycle before we can reuse it.
+     */
+    __asm__ __volatile__ ("nop");
```

Comment:
The "ret" instruction that the function need to execute takes 4 to 5 cycles. The CPU takes three clock cycles, minimum to enter in any "isr"; one cycle to finish the ongoing instruction and two cycles to store the program counter to the stack. These means that the "nop" is not necessary.

Reply:
OK. It might worth pointing out that this function is highly likely got get inlined when compiling with `LTO=1`. The next steps in the boot process are `avr8_stdio_init()` (only when stdio is used) and `periph_init()`. Those are again only called from `cpu_init()`, so inlining them with `LTO=1` is again likely.

I don't think that it would cause in issue for either `avr8_stdio_init()` or `periph()`. But it might be worth to keep the comment here to ease debugging, if I'm wrong.

Also: Since `clk_init()` generates a symbol, it is better to "namespace" it. E.g. `cpu_clk_init()` or `avr8_clk_init()`.

Reply:
I complete remove the NOP instructions, It were not originally designed by me. I've been running tests and didn't see instability. I took a look at the resulting "lst" file, snip below. The below is showed in the inline documentation file from the file that define the macro at `/usr/lib/avr/include/avr/xmega.h`. I'm assuming this construction is fine since there is a example how to use and the version without NOP is stabe.

```
    _PROTECTED_WRITE(CLK.PSCTRL, clk_scale | bus_scale);
    2232:	98 ed       	ldi	r25, 0xD8	; 216
    2234:	20 e0       	ldi	r18, 0x00	; 0
    2236:	30 e0       	ldi	r19, 0x00	; 0
    2238:	94 bf       	out	0x34, r25	; 52
    223a:	20 93 41 00 	sts	0x0041, r18	; 0x800041 <__TEXT_REGION_LENGTH__+0x7e0041>
    atxmega_set_prescaler(CPU_ATXMEGA_CLK_SCALE_INIT,
                          CPU_ATXMEGA_BUS_SCALE_INIT);

    _PROTECTED_WRITE(CLK.CTRL, CLK_SCLKSEL_RC32M_gc);
    223e:	94 bf       	out	0x34, r25	; 52
    2240:	80 93 40 00 	sts	0x0040, r24	; 0x800040 <__TEXT_REGION_LENGTH__+0x7e0040>
    2244:	08 95       	ret
```